src/transformStateWithClones.js

```diff
@@ -8,6 +8,29 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete currentState[key]);
+        break;
+      default:
+        break;
+    }
+
+    stateHistory.push({ ...currentState });
```

Comment:
It is more efficient to push the copy of the state to the `stateHistory` array outside of the `switch` block, but still within the `forEach` loop, to avoid repeating this action in every case.